File: Java/328-OddEvenLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null)  return head;
        ListNode preHead1 = null, preHead2 = null;
        ListNode evenHead = head;
        ListNode oddHead = head.next;
        ListNode even = evenHead;
        ListNode odd = oddHead;
        while (even.next != null && odd.next != null){
            even.next = odd.next;
            even = even.next;
            odd.next = even.next;
            odd = odd.next;
        }
        even.next = oddHead;
        return evenHead;
    }
}

File: cpp/328-OddEvenLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 //      c1 c2
 //Given 1->2->3->4->5->NULL,
 //return 1->3->5->2->4->NULL.
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }
        ListNode* head2 = head->next;
        ListNode* odd = head, *even = head2;
        while (odd->next && even->next) {
            odd->next = odd->next->next;
            even->next = even->next->next;
            odd = odd->next;
            even = even->next;
        }
        odd->next = head2;
        return head;
    }
};
